File: de_lekbak/backend/de_lekbak_backend/schemas/viral_cve.py
from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, Field


class SourceType(StrEnum):
    reddit = "reddit"
    mastodon = "mastodon"
    hacker_news = "hacker_news"


class SourceLink(BaseModel):
    source_type: SourceType
    title: str
    url: str
    observed_at: datetime | None = None


class NvdEnrichment(BaseModel):
    severity: str | None = None
    description: str | None = None
    affected_vendors: list[str] = Field(default_factory=list)
    affected_products: list[str] = Field(default_factory=list)
    published_at: datetime | None = None
    modified_at: datetime | None = None


class ViralCveItem(BaseModel):
    cve_id: str
    virality_score: float
    mention_count: int
    distinct_source_count: int
    source_types: list[SourceType]
    first_seen_at: datetime | None = None
    last_seen_at: datetime | None = None
    representative_links: list[SourceLink] = Field(default_factory=list)
    nvd: NvdEnrichment | None = None


class ViralCveRankingResponse(BaseModel):
    items: list[ViralCveItem]
    last_refreshed_at: datetime | None = None
    is_stale: bool = False


class ViralCveRefreshResponse(BaseModel):
    rankings: ViralCveRankingResponse
    message: str

File: de_lekbak/frontend/src/types/viralCve.ts
export type SourceType = "reddit" | "mastodon" | "hacker_news";

export interface SourceLink {
  source_type: SourceType;
  title: string;
  url: string;
  observed_at: string | null;
}

export interface NvdEnrichment {
  severity: string | null;
  description: string | null;
  affected_vendors: string[];
  affected_products: string[];
  published_at: string | null;
  modified_at: string | null;
}

export interface ViralCveItem {
  cve_id: string;
  virality_score: number;
  mention_count: number;
  distinct_source_count: number;
  source_types: SourceType[];
  first_seen_at: string | null;
  last_seen_at: string | null;
  representative_links: SourceLink[];
  nvd: NvdEnrichment | null;
}

export interface ViralCveRankingResponse {
  items: ViralCveItem[];
  last_refreshed_at: string | null;
  is_stale: boolean;
}

export interface ViralCveRefreshResponse {
  rankings: ViralCveRankingResponse;
  message: string;
}
